File: routes/gamingGen.mjs
import express from "express";
import puppeteer from "puppeteer-extra";
import StealthPlugin from "puppeteer-extra-plugin-stealth";

puppeteer.use(StealthPlugin());

const router = express.Router();

async function scrapePage(browser, url) {
  const page = await browser.newPage();
  try {
    console.log(`Navigating to ${url}`);
    await page.goto(url, { waitUntil: "networkidle2", timeout: 90000 });
    
    // Cloudflare bypass wait
    try {
        await page.waitForSelector("li.product", { timeout: 15000 });
    } catch (e) {
        console.log("Product selector not found, might be blocked or empty page.");
    }

    const products = await page.evaluate(() => {
      // Selector fallbacks for container
      let items = Array.from(document.querySelectorAll("li.product"));
      if (items.length === 0) items = Array.from(document.querySelectorAll(".product"));
      if (items.length === 0) items = Array.from(document.querySelectorAll(".pc-specs-list-container")); // Alternate layout

      return items.map(item => {
        // Selector fallbacks for title
        const titleSelectors = [".pc-specs-title", ".woocommerce-loop-product__title", "h2", "h3", ".product-title"];
        let name = "N/A";
        for (const sel of titleSelectors) {
          const el = item.querySelector(sel);
          if (el && el.textContent.trim()) {
            name = el.textContent.trim();
            break;
          }
        }
        const link = item.querySelector("a")?.href || null;
        const imgEl = item.querySelector("img");
        const image = imgEl ? (imgEl.getAttribute("data-lazy-src") || imgEl.getAttribute("data-src") || imgEl.src) : null;
        
        // Fiyat parse
        const priceElement = item.querySelector(".price");
        let priceText = "";
        if (priceElement) {
            const ins = priceElement.querySelector("ins");
            priceText = ins ? ins.textContent : priceElement.textContent;
        }
        const match = priceText.replace(/[^\d,]/g, "").replace(",", ".");
        const price = parseFloat(match) || 0;

        // Specs (Smart Finder)
        const specs = { CPU: "N/A", Motherboard: "N/A", GPU: "N/A", Ram: "N/A", Storage: "N/A" };
        const specItems = Array.from(item.querySelectorAll(".pc-specs-list li")).map(li => li.textContent.trim());
        
        const find = (kws) => specItems.find(x => kws.some(k => x.toLowerCase().includes(k.toLowerCase()))) || "N/A";
        
        specs.CPU = find(["islemci", "cpu", "ryzen", "core", "intel", "amd", "i3", "i5", "i7", "i9"]);
        specs.Motherboard = find(["anakart", "mb", "b450", "b550", "a520", "h610", "b650", "a620", "b760", "z790", "b660", "x670"]);
        specs.GPU = find(["rtx", "gtx", "rx ", "arc", "radeon", "ekran"]);
        specs.Ram = find(["mhz", "ram", "ddr", "cl"]);
        specs.Storage = find(["ssd", "m.2", "nvme", "tb"]);

        // Name fallback for specs
        if (specs.CPU === "N/A" && name !== "N/A") {
            const cpuMatch = name.match(/(INTEL[\w\s]+|AMD[\w\s]+|INTE\s+U\d[\w\s]+)/i);
            if (cpuMatch) specs.CPU = cpuMatch[1].trim();
        }
        if (specs.GPU === "N/A" && name !== "N/A") {
            const gpuMatch = name.match(/((?:RTX|GTX|RX|ARC|RADEON)\s*\d+[\w\s]*)/i);
            if (gpuMatch) specs.GPU = gpuMatch[1].trim();
        }

        return { name, price, image, link, specs, store: "gamingGen" };
      });
    });

    return products;
  } catch (error) {
    console.error(`Error scraping ${url}:`, error);
    return [];
  } finally {
    await page.close();
  }
}

export async function scrapeAllPages() {
  const browser = await puppeteer.launch({
    headless: true,
    args: ["--no-sandbox", "--disable-setuid-sandbox"],
  });

  try {
    const baseUrl = "https://www.gaming.gen.tr/kategori/hazir-sistemler/";
    const page = await browser.newPage();
    await page.setUserAgent("Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/123.0.0.0 Safari/537.36");
    await page.goto(baseUrl, { waitUntil: "networkidle2", timeout: 90000 });
    
    // Wait for content or bypass
    try {
        await page.waitForSelector("li.product", { timeout: 20000 });
    } catch(e) {
        console.log("Initial page load timed out or blocked.");
    }
    
    // Find total pages
    const totalPages = await page.evaluate(() => {
        const pageLinks = Array.from(document.querySelectorAll(".page-numbers:not(.next)"));
        const pageNumbers = pageLinks
            .map(el => {
                const text = el.textContent.trim().replace(/[^\d]/g, "");
                return parseInt(text, 10);
            })
            .filter(n => !isNaN(n));
        return pageNumbers.length ? Math.max(...pageNumbers) : 1;
    });
    await page.close();

    console.log(`Found ${totalPages} pages for GamingGen`);

    const results = [];
    const pagesToScrape = Math.min(totalPages, 60); 

    for (let i = 1; i <= pagesToScrape; i++) {
        const url = i === 1 ? baseUrl : `${baseUrl}page/${i}/`;
        const pageProducts = await scrapePage(browser, url);
        if (pageProducts.length > 0) {
            results.push(...pageProducts);
        } else if (i > 1) {
            // If we get 0 products on a subpage, maybe we reached the end or got blocked
            console.log(`No products found on page ${i}, stopping.`);
            break;
        }
        // Random delay to avoid bot detection
        await new Promise(r => setTimeout(r, 2000 + Math.random() * 3000));
    }

    return results;
  } finally {
    await browser.close();
  }
}

router.get("/", async (req, res) => {
  try {
    const products = await scrapeAllPages();
    res.json(products);
  } catch (err) {
    res.status(500).json({ error: err.message });
  }
});

export default router;

File: python_backend/gaminggen.py
"""
GamingGen scraper - StealthyFetcher
- wait_selector: li.product (pagination degil)
- Concurrent page fetching with Semaphore(3)
"""
import asyncio
import re
from scrapling.fetchers import StealthyFetcher

STORE = "gamingGen"
BASE_URL = "https://www.gaming.gen.tr/kategori/hazir-sistemler/"
PRODUCT_SEL = "li.product"


def _parse_page_products(page) -> list[dict]:
    products = []
    for item in page.css(PRODUCT_SEL):
        title_el = item.css(".pc-specs-title").first
        name = title_el.text.strip() if title_el else "N/A"
        a_el = item.css("a").first
        link = a_el.attrib.get("href") if a_el else None
        img_el = item.css("img").first
        image = img_el.attrib.get("src") if img_el else None
        price = 0.0
        price_el = item.css(".price").first
        if price_el:
            ins_el = price_el.css("ins").first
            price_text = ins_el.text if ins_el else price_el.text
            raw = re.sub(r"[^\d,]", "", price_text).replace(",", ".")
            try:
                price = float(raw)
            except ValueError:
                price = 0.0
        spec_items = [li.text.strip() for li in item.css(".pc-specs-list li")]
        def find(*kws):
            return next((x for x in spec_items if any(k.lower() in x.lower() for k in kws)), "N/A")
        specs = {
            "CPU": find("islemci", "cpu", "ryzen", "core", "intel", "amd"),
            "Motherboard": find("anakart", "mb"),
            "GPU": find("rtx", "gtx", "rx ", "arc"),
            "RAM": find("mhz", "ram", "ddr", "cl"),
            "Storage": find("ssd", "m.2", "nvme", "tb"),
        }
        products.append({"name": name, "price": price, "image": image, "url": link, "store": STORE, "specs": specs})
    return products


async def _fetch_page(url: str) -> list[dict]:
    page = await StealthyFetcher.async_fetch(url, headless=True, wait_until="domcontentloaded",
                                              timeout=60000, wait_selector=PRODUCT_SEL, wait_selector_state="attached")
    return _parse_page_products(page)


async def scrape_all_pages_async() -> list[dict]:
    print(f"[GamingGen] Ilk sayfa: {BASE_URL}", flush=True)
    try:
        first_page = await StealthyFetcher.async_fetch(BASE_URL, headless=True, wait_until="domcontentloaded",
                                                        timeout=60000, wait_selector=PRODUCT_SEL, wait_selector_state="attached")
    except Exception as e:
        print(f"[GamingGen] Ilk sayfa hatasi: {e}", flush=True)
        return []

    try:
        nums = [int(re.sub(r"[^\d]", "", el.text.strip())) for el in first_page.css(".page-numbers:not(.next)") if re.sub(r"[^\d]", "", el.text.strip()).isdigit()]
        total_pages = min(max(nums), 60) if nums else 1
    except Exception:
        total_pages = 1

    print(f"[GamingGen] {total_pages} sayfa", flush=True)
    all_products = _parse_page_products(first_page)

    if total_pages > 1:
        sem = asyncio.Semaphore(3)
        async def fetch_n(n):
            async with sem:
                for attempt in range(3):
                    try:
                        return await _fetch_page(f"{BASE_URL}page/{n}/")
                    except Exception as e:
                        print(f"[GamingGen] Sayfa {n} hata (deneme {attempt+1}): {e}", flush=True)
                        await asyncio.sleep(3)
                return []
        results = await asyncio.gather(*[fetch_n(i) for i in range(2, total_pages + 1)], return_exceptions=True)
        for r in results:
            if isinstance(r, list):
                all_products.extend(r)

    print(f"[GamingGen] Toplam {len(all_products)} urun", flush=True)
    return all_products


def scrape_all_pages() -> list[dict]:
    return asyncio.run(scrape_all_pages_async())


if __name__ == "__main__":
    import json
    products = scrape_all_pages()
    with open("gaminggen_test.json", "w", encoding="utf-8") as f:
        json.dump(products, f, ensure_ascii=False, indent=2)
    print("gaminggen_test.json kaydedildi")
